liip/LiipImagineBundle
# [Console] Lazy commands, separate IO concerns, add forced resolve to command, and prittify and detail output

| Q | A
| --- | ---
| Branch? | 2.0 
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets | 
| License | MIT
| Doc PR | 

This is the `2.0` version of #967. Included are commands with the following new features:

- Use of commands as a service
- Removal of "ContainerAwareInterface" from commands (instead, the container is used to pass required dependencies)
- Updated styling that includes colored output, as well as the ability to disable colors and print machine-readable output
- Separation of concerns between the commands, the IO/style, etc
- Enable comparable output for the remove cache command

The use of explicit dependencies (via the constructor), as well as the separation of concerns, results in relatively simple commands, despite the complex behavior and output styling introduced. See:

- [ResolveCacheCommand.php](https://github.com/robfrawley/LiipImagineBundle/blob/c136b8fbad60ff31ae323b8f221c1e322f837872/Command/ResolveCacheCommand.php)
- [RemoveCacheCommand.php](https://github.com/robfrawley/LiipImagineBundle/blob/c136b8fbad60ff31ae323b8f221c1e322f837872/Command/RemoveCacheCommand.php)

It is also worth noting that the bulk of the PR size is new and updated unit tests. The new command output for the cache resolve command looks like the following:

![imagine-cli-resolve_cached-resolved-failed_normal](https://user-images.githubusercontent.com/3967713/29899460-a86ef52c-8db9-11e7-9f72-d526345bae4f.png)
![imagine-cli-resolve_cached-resolved-failed_no-colors](https://user-images.githubusercontent.com/3967713/29899464-aeea9ae6-8db9-11e7-8850-ddeb2eb02f86.png)
![imagine-cli-resolve_cached-resolved-failed_as-script](https://user-images.githubusercontent.com/3967713/29899468-b1607a0c-8db9-11e7-8b0b-9f1e97084350.png)
![imagine-cli-resolve_cached-resolved_normal](https://user-images.githubusercontent.com/3967713/29899472-b64755fe-8db9-11e7-95d3-0067876018d3.png)
![imagine-cli-resolve_failed_all](https://user-images.githubusercontent.com/3967713/29899474-b8926de4-8db9-11e7-95af-b0ddf5403616.png)
